HDDS-2344. Add immutable entries in to the DoubleBuffer for Volume requests.

bharatviswa504 opened the pull request:
…quests.

## What changes were proposed in this pull request?

OMVolumeCreateRequest.java L159:

omClientResponse =
 new OMVolumeCreateResponse(omVolumeArgs,volumeList, omResponse.build());
 

We add this to double-buffer, and double-buffer flushThread which is running in the background when picks up, converts to protoBuf and to ByteArray and write to rocksDB tables. So, during this conversion(This conversion will be done without any lock acquire), if any other request changes internal structure(like acls list) of OmVolumeArgs we might get ConcurrentModificationException.


## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-2344



## How was this patch tested?

Ran TestOzoneRpcClient which tests this code path and also added a new test for clone method.

dineshchitlangia commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmOzoneAclMap.java:
```diff
@@ -298,4 +303,27 @@ public static OmOzoneAclMap ozoneAclGetFromProtobuf(
+    defaultList.addAll(defaultAclList);
+
+    return new OmOzoneAclMap(defaultList, accessMap);
+  }
```
OmOzoneAclMap.clone() does not call super.clone(). Thus the correct object type is not guaranteed here and violates the contract for clone(). Most likely this is the findbug issue listed in CI run. To fix this, you must call the super.clone()

bharatviswa504 replied:
I thought it is the final class and did like that. I think this should be marked as final like other classes like OMVolumeArgs. (As these classes represent proto structure data). And also as for few parameters, I need a deep copy, so I changed like this. Let me know what do you think.

dineshchitlangia replied:
Given the use case and avoid shallow copy, it is better to make it final.

dineshchitlangia commented on hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/helpers/TestOmVolumeArgs.java:
```diff
@@ -0,0 +1,86 @@
+/**
+ *
+ */
+public class TestOmVolumeArgs {
```
Empty javadoc, let's use it or scrap it :)

bharatviswa504 replied:
Updated it.

dineshchitlangia commented on hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/helpers/TestOmVolumeArgs.java:
```diff
@@ -29,7 +29,7 @@
 
 /**
- *
+ * Class used to Test OmVolumeArgs.
```
NIT: `Test` -> `test`

bharatviswa504 replied:
Done.

dineshchitlangia commented:
Other test failures seem unrelated.

xiaoyuyao commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/volume/OMVolumeSetOwnerRequest.java:
```diff
@@ -171,7 +171,8 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
       omClientResponse = new OMVolumeSetOwnerResponse(oldOwner,
-          oldOwnerVolumeList, newOwnerVolumeList, omVolumeArgs,
+          oldOwnerVolumeList, newOwnerVolumeList,
+          (OmVolumeArgs) omVolumeArgs.clone(),
```
It is not clear to me why an extra clone of omVolumeArgs is needed here. omVolumeArgs is just created a few line above in Line 129. Since it is not an argument provided by the caller and not handed over to other clients, can you post stack tracing showing concurrent modification exception?

bharatviswa504 replied:
https://issues.apache.org/jira/browse/HDDS-2322
The issue is seen with Key Operations, I think we might see the same issue with Other Ops too. Let's take a case where RemoveAcl submitted OMVolumeArgs to flushThread and other thread like removeAcl is updating same OmVolumeArgs on this. Then when we might see ConcurrentModificationException. So to cover these kinds of scenario's updated all the places where we submit the response to doubleBuffer flush threads.

dineshchitlangia replied:
@xiaoyuyao does this address your concern?

xiaoyuyao replied:
bq. other thread which is performing removeAcl is updating the same OmVolumeArgs on this(Like we have a list and map internally for acls)

We've already done defensive copy for get Java object of Volume/Bucket/Key from DB and converted the client provided acl from protobuf when creating OMxxxAclRequest. I think we should trace down the real cause of the CME here. Generally, I agree making object immutable is a good thing. However, unnecessary clone object has performance implications  that we should be cautious. 

bq. when flush happens we might flush these entries to rocksdb when some other operation is changing the same entry which might be still submitted/not submitted to doublebuffer flush thread.

omVolumeArgs/KeyInfo/BucketInfo are created per request from DB like below:
     omVolumeArgs = omMetadataManager.getVolumeTable().get(dbVolumeKey);

Can you elaborate on the operation that could change the per request created volume/bucket/key object?

bharatviswa504 replied:
Yes but with cache enabled, every time it is not a new object.

bharatviswa504 replied:
Thank You @xiaoyuyao for the offline discussion. Posted a patch to make immutable value from get() API when it is returned from cache. (In this updated only for Bucket and Volume requests) Updated bucket request also in this, as that got checked in as part of HDDS-2333. Please share your view on this.

xiaoyuyao commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmBucketInfo.java:
```diff
@@ -366,7 +367,6 @@ public static OmBucketInfo getFromProtobuf(BucketInfo bucketInfo) {
     return obib.build();
   }
 
-  @Override
```
Why do we remove the @Override for equals?

bharatviswa504 replied:
Sorry mistakenly removed it. Reverted it back.

xiaoyuyao commented:
+1 pending CI. Thanks @bharatviswa504  for the offline discussion and update.

bharatviswa504 commented:
Thank You @xiaoyuyao for the review.
Test failure in OzoneNativeAuthorizer is failing now, as we return new Objects. This object is previously passing because we used to have references and used to update the same references. But now we changed to return copy object, and the Old Api's update DB not cache. The test is failing.  I have disabled this test for now. I will open a new jira to fix this.

dineshchitlangia commented:
> Thank You @xiaoyuyao for the review.
> Test failure in OzoneNativeAuthorizer is failing now, as we return new Objects. This object is previously passing because we used to have references and used to update the same references. But now we changed to return copy object, and the Old Api's update DB not cache. The test is failing. I have disabled this test for now. I will open a new jira to fix this.

@bharatviswa504 Since the test is failing due to the changes in this PR, I think it would be great if we can fix the test here as we already know the cause. If this jira is a blocker and need to commit urgently then it is even better to fix it in this PR.

bharatviswa504 commented:
> > Thank You @xiaoyuyao for the review.
> > Test failure in OzoneNativeAuthorizer is failing now, as we return new Objects. This object is previously passing because we used to have references and used to update the same references. But now we changed to return copy object, and the Old Api's update DB not cache. The test is failing. I have disabled this test for now. I will open a new jira to fix this.
> 
> @bharatviswa504 Since the test is failing due to the changes in this PR, I think it would be great if we can fix the test here as we already know the cause. If this jira is a blocker and need to commit urgently then it is even better to fix it in this PR.

Thank You, Dinesh for the comments. I have fixed the test case.

bharatviswa504 commented:
/retest

dineshchitlangia commented:
@bharatviswa504 Thanks for quick turn around. I verified in my local and seems fine. +1 pending CI, will merge once done.

dineshchitlangia commented:
Thanks @xiaoyuyao for reviews and @bharatviswa504 for contribution. This has been merged to master.
